packages/dd-trace/src/appsec/rule_manager.js

```diff
@@ -188,67 +111,28 @@ function concatArrays (files) {
   return [...files.values()].flat()
 }
 
-/*
-  ASM_DATA Merge strategy:
-  The merge should be based on the id and type. For any duplicate items, the longer expiration should be taken.
-  As a result, multiple Rule Data may use the same DATA_ID and DATA_TYPE. In this case, all values are considered part
-  of a set and are merged. For instance, a denylist customized by environment may use a global Rule Data for all
-  environments and a Rule Data per environment
-*/
+function extractErrors (obj) {
+  if (typeof obj !== 'object' || obj === null) return null
 
-function mergeRulesData (files) {
-  const mergedRulesData = new Map()
-  for (const [, file] of files) {
-    for (const ruleData of file) {
-      const key = `${ruleData.id}+${ruleData.type}`
-      if (mergedRulesData.has(key)) {
-        const existingRulesData = mergedRulesData.get(key)
-        ruleData.data.reduce(rulesReducer, existingRulesData.data)
-      } else {
-        mergedRulesData.set(key, copyRulesData(ruleData))
+  const result = {}
+
+  for (const [key, value] of Object.entries(obj)) {
+    if (key === 'error' || key === 'errors') {
+      result[key] = value
+    } else if (typeof value === 'object' && value !== null) {
+      const child = extractErrors(value)
+      if (child && Object.keys(child).length > 0) {
+        result[key] = child
       }
     }
   }
-  return [...mergedRulesData.values()]
-}
-
-function rulesReducer (existingEntries, rulesDataEntry) {
-  const existingEntry = existingEntries.find((entry) => entry.value === rulesDataEntry.value)
-  if (existingEntry && !('expiration' in existingEntry)) return existingEntries
-  if (existingEntry && 'expiration' in rulesDataEntry && rulesDataEntry.expiration > existingEntry.expiration) {
-    existingEntry.expiration = rulesDataEntry.expiration
-  } else if (existingEntry && !('expiration' in rulesDataEntry)) {
-    delete existingEntry.expiration
-  } else if (!existingEntry) {
-    existingEntries.push({ ...rulesDataEntry })
-  }
-  return existingEntries
-}
 
-function copyRulesData (rulesData) {
-  const copy = { ...rulesData }
-  if (copy.data) {
-    const data = []
-    copy.data.forEach(item => {
-      data.push({ ...item })
-    })
-    copy.data = data
-  }
-  return copy
+  return Object.keys(result).length > 0 ? result : null
```

Comment:
Instead of checking for results keys, I would just use a boolean that I set to true in case the loop happened to match something, since calling `Object.keys()` is quite expensive to just get that information.

In general, is the object really arbitrary that we have to loop over it completely in an unknown way?

Reply:
Nope, the object is not really arbitrary, it has a defined schema (at least, for this version of libddwaf).